### ISSUE TITLE
Automatic notification for recurred invoices during cron task is not sending the invoice pdf

### ISSUE THREAD

**Issue by Cirillo007** [Dolibarr/dolibarr#32069]
### Bug

Hello everyone,

when using the cron and the creation of new recurred invoices, using the automatic notification of new invoice for third party, the mail send talks about an invoice that is not validated (PROV) and there is no pdf attached to the mail.

Discovering the log, it seems that the sending process come before the validation process of invoice. That's why there is no pdf and name is PROV.

### Dolibarr Version

19.0.3

### Environment PHP

8.2

### Steps to reproduce the behavior and expected behavior

For a client, I've decided to notify him every time there is a new invoice validated. 
So I've created the third party as follow:
<img width="690" alt="image" src="https://github.com/user-attachments/assets/59ae9f9e-26dc-4300-8d1e-dff666286f1d">

I create a recur invoice model.

In email I've selected the right format FCAuto as follow:
<img width="543" alt="image" src="https://github.com/user-attachments/assets/c1f62f49-cc5e-4739-b119-d8f617a0124c">

In automatic notification I've selected also the right preset:
<img width="547" alt="image" src="https://github.com/user-attachments/assets/35b96e33-45a7-48ec-b796-3be45d8e5686">

The cron worked well and created the new invoice as it should do so till this point it's all ok
<img width="547" alt="image" src="https://github.com/user-attachments/assets/9e3f8ad0-3052-439a-a2f0-ea47b9bb6001">

<img width="544" alt="image" src="https://github.com/user-attachments/assets/972a9bf6-2fa9-4644-90d9-aa670b675afd">

But in the email received, it says:
<img width="555" alt="image" src="https://github.com/user-attachments/assets/b4a9b5e1-f8f4-4c56-bd81-af49c484f5a7">

So first thing wrong:
- the name of the invoice is PROV47 that is an invoice not validated.
- the pdf doesn't exists

In the invoice module all is correct to me 
<img width="547" alt="image" src="https://github.com/user-attachments/assets/bc476583-aca4-4dcb-8b69-7c19dfd0d4e1">

When I saw the lof, I saw that mail are send before validating the invoice.
In the log I see this :
CMailFile::CMailfile: charset=UTF-8 from=
CMailFile::CMailfile: subject= Votre facture (PROV44)
CMailFile::CMailfile: sendmode=smtps addr_bcc=, replyto=
HookManager::initHooks Loading hooks: context=mail-path=/removeconf/class/actions_removeconf.class.php
CMailFile::sendfile: sendMsg, HOST=ssl://ssl0.ovh.net, PORT=465
CMailFile::sendfile: mail end success

and AFTER I see this:
commonGenerateDocument modele=sponge outputlangs->defaultlang=fr_FR

So that could tell that the mail is send before the validation process.

Is it enough to understand this bug ?

I had also the same issue while validating memberships where a new member received the mail of the new membership saying that it was enclosed the invoice, but without the invoice. Maybe I can reproduce the issue in another article ?

Thank you,

**Comment by homer8173** [comment]
Yes it's a problem of conception, @eldy how do you want to fix it?

**Comment by Cirillo007** [comment]
> Yes it's a problem of conception, [@eldy](https://github.com/eldy) how do you want to fix it?

Do you know how we can make this but by handled ?

**Comment by JoeBiboche** [comment]
Same behaviour in 20.0.1 with _Notification on events_ module. Waiting for this fix in order to help automating invoice notifications with attachment (interested for validated and paid events).

**Comment by homer8173** [comment]
Hello @eldy,
one more time I had this subject in meeting.
Invoicing automation is a key feature and the position of the triggers make this impossible to accomplish.
You can call me if you wanna talk about it.

**Comment by eldy** [comment]
> > Yes it's a problem of conception, [@eldy](https://github.com/eldy) how do you want to fix it?
> 
> Do you know how we can make this but by handled ?

I don't see any conception trouble;

By definition, a trigger is run inside a transaction, not before validation, and not after, but inside (because a trigger must be able to cancel a transaction)
When using the notification module, it is a trigger. The trigger receive both the old ref and new ref.
If the ref is wrong in your email, it may be because the string in template is __REF__ instead of __NEWREF__ (for a validation action, the 2 are defined to 2 different values).

About the pdf, the pdf is created once invoice is validated (because we want a pdf of the invoice once all validation steps are done). So, currently, the notification module, that send email before end, is not able to send  attachment.

If you need to send email with attachement of pdf generated after the pdf has been generated, this need another module than the module Notification in core. And this module must not use trigger to ruj its task (exception if action can be done inside validation). If module need to interact after  the validation and generation of pdf, the module must use hook, for example the hook afterCreationOfRecurringInvoice.

**Comment by homer8173** [comment]
With an external module, I agree you can do whatever you want, but here we're talking about the core.

So you confirm that Dolibarr will never comply with the basic promise it makes to send the invoices it generates automatically?

**Comment by Cirillo007** [comment]
I understand Eldy your point of view.
Can we agree that seeing Facture client validée seems to the user that as soon as an invoice as been validated we can trigger an email that is send.
Do you thing that we can trigger in other way to trigger after pdf creation ?

**Comment by daxitsolutions** [comment]
This is still happening in v22.0.4.

Using __NEWREF__ fixes the reference in the notification, but the PDF is still missing, which is unfortunate.

At the moment, the only workaround we found is to send the customer a direct download link to the document, e.g.:
https://xxxx/htdocs/document.php?hashp=8S7LyYR2W3ME987xb5zo9uWq6Sx1Jr1vm4N
(we generate it via __DIRECTDOWNLOAD_URL_ORDER__, for example).

Also, this looks like a regression: in older versions we had a checkbox allowing us to choose who receives the notification. That workflow is no longer possible.

**Comment by pixodeo** [comment]
Traced this. On the maintained branches (21.0, 22.0, 23.0), FactureRec::createRecurringInvoices() validates the invoice and then generates the PDF, but it does not send the invoice by email itself. So the only email that can go out is the one from the Notification module trigger on BILL_VALIDATE, and that trigger fires inside validate() - i.e. before generateDocument() runs in the cron. At that point last_main_doc/the PDF does not exist yet, so the notification is sent without the PDF attached (and, depending on timing, with the provisional ref), exactly as you observed.

On develop this is already fixed, but by a feature rather than a small patch: commit 8a8a7f4e954 "NEW #25829 Automatically send the invoice generated from a template (#36967)" adds an explicit email send inside createRecurringInvoices, placed after generateDocument(), so the PDF is attached and the ref is the validated one. That block is not present on 22.0/23.0.

So: the proper resolution is that feature (#36967 / #25829), which lives on develop and will ship in the next major. Backporting it to 21.0-23.0 would mean backporting a NEW feature onto maintenance branches, which is a maintainer call - I did not want to do that unilaterally. Flagging with the exact cause and the commit so it can be decided.
